File: webrtc-jni/src/main/cpp/src/media/video/VideoCapture.cpp
/*
 * Copyright 2021 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "media/video/VideoCapture.h"
#include "Exception.h"

#include "api/video/i420_buffer.h"
#include "modules/video_capture/video_capture_factory.h"

namespace jni
{
	VideoCapture::VideoCapture() :
		captureModule(nullptr)
	{
		capability.width = static_cast<int32_t>(1280);
		capability.height = static_cast<int32_t>(720);
		capability.maxFPS = static_cast<int32_t>(30);
	}

	VideoCapture::~VideoCapture()
	{
		destroy();
	}

	void VideoCapture::setDevice(const avdev::DevicePtr & device)
	{
		this->device = device;
	}

	void VideoCapture::setVideoCaptureCapability(const webrtc::VideoCaptureCapability & capability)
	{
		this->capability = capability;
	}

	void VideoCapture::setVideoSink(std::unique_ptr<rtc::VideoSinkInterface<webrtc::VideoFrame>> sink)
	{
		this->sink = std::move(sink);
	}

	void VideoCapture::start()
	{
		if (!device) {
			throw new Exception("Video device must be set");
		}

		std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> info(webrtc::VideoCaptureFactory::CreateDeviceInfo());

		if (!info) {
			throw new Exception("Create video DeviceInfo failed");
		}

		uint32_t num = info->NumberOfDevices();

		if (num < 1) {
			throw new Exception("No video capture devices available");
		}
		
		std::string devUid;
		const uint32_t size = webrtc::kVideoCaptureDeviceNameLength;

		for (uint32_t i = 0; i < num; ++i) {
			char name[size] = { 0 };
			char guid[size] = { 0 };

			int32_t ret = info->GetDeviceName(i, name, size, guid, size);

			if (ret != 0) {
				RTC_LOG(LS_WARNING) << "Get video capture device name failed";
				continue;
			}

			if (device->getName().compare(name) == 0) {
				devUid = guid;
				break;
			}
		}

		if (devUid.empty()) {
			throw new Exception("Device %s not found", device->getName().c_str());
		}

		captureModule = webrtc::VideoCaptureFactory::Create(devUid.c_str());

		if (!captureModule) {
			throw new Exception("Create VideoCaptureModule for UID %s failed", devUid.c_str());
		}

		captureModule->RegisterCaptureDataCallback(std::move(sink).release());

		capability.videoType = webrtc::VideoType::kI420;

		if (captureModule->StartCapture(capability) != 0) {
			destroy();

			throw new Exception("Start video capture for UID %s failed", devUid.c_str());
		}

		if (!captureModule || !captureModule->CaptureStarted()) {
			throw new Exception("Start video capture failed");
		}
	}

	void VideoCapture::stop()
	{
		destroy();
	}

	void VideoCapture::destroy()
	{
		if (!captureModule) {
			return;
		}
		if (captureModule->CaptureStarted()) {
			captureModule->StopCapture();
		}

		captureModule->DeRegisterCaptureDataCallback();
		captureModule = nullptr;
	}
}

File: webrtc/src/main/java/dev/onvoid/webrtc/media/video/VideoCapture.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc.media.video;

import dev.onvoid.webrtc.internal.NativeObject;

/**
 * A class representing a video capture device that can be controlled to capture video.
 * Extends NativeObject to interact with native code implementations.
 * <pre>
 * VideoCapture capture = new VideoCapture();
 * capture.setVideoCaptureDevice(device);
 * capture.setVideoCaptureCapability(capability);
 * capture.setVideoSink(sink);
 * capture.start();
 * // ... consume frames via sink ...
 * capture.stop();
 * capture.dispose();
 * </pre>
 * Notes:
 * <li>Always call dispose() to release native resources.</li>
 * <li>Methods are not guaranteed to be thread-safe unless externally synchronized.</li>
 *
 * @author Alex Andres
 */
public class VideoCapture extends NativeObject {

	/**
	 * Constructs a new VideoCapture and initializes underlying native resources.
	 * Initialization failures are typically surfaced as runtime exceptions
	 * originating from native code.
	 */
	public VideoCapture() {
		initialize();
	}

	/**
	 * Selects the physical (or virtual) video input device to capture from.
	 * <p>
	 * Constraints / Lifecycle:
	 * <li>Must be invoked before {@link #start()}.</li>
	 * <li>Re-setting the device after capture has started may require an
	 *     internal restart (implementation-dependent).</li>
	 *
	 * @param device Non-null device descriptor to bind. Passing {@code null}
	 *               is invalid and may raise a {@link NullPointerException}.
	 *
	 * @throws IllegalStateException if called after disposal.
	 */
	public native void setVideoCaptureDevice(VideoDevice device);

	/**
	 * Defines desired capture parameters (resolution, frame rate, pixel format, etc.).
	 * <p>
	 * Constraints:
	 * <li>Should be called after selecting the device and before {@link #start()}.</li>
	 * <li>Some capabilities may be adjusted (e.g., negotiated to the nearest supported values).</li>
	 *
	 * @param capability Desired capture capability (must be non-null).
	 *
	 * @throws IllegalArgumentException if unsupported or invalid.
	 * @throws IllegalStateException    if called after disposal.
	 */
	public native void setVideoCaptureCapability(VideoCaptureCapability capability);

	/**
	 * Registers (or replaces) the sink that will receive decoded/raw video frames.
	 * <p>
	 * Behavior:
	 * <li>May be called before or after {@link #start()}.</li>
	 * <li>Passing {@code null} (if supported) detaches the current sink and
	 *     frames will be dropped until a new sink is set.</li>
	 *
	 * @param sink The consumer of captured frames.
	 *
	 * @throws IllegalStateException if called after disposal.
	 */
	public native void setVideoSink(VideoTrackSink sink);

	/**
	 * Begins asynchronous frame capture and delivery to the configured sink.
	 * <p>
	 * Idempotency:
	 * Calling start() while already started should be a no-op (implementation-dependent).
	 *
	 * @throws IllegalStateException if prerequisites (device/capability) are missing
	 *                               or the instance is disposed.
	 */
	public native void start();

	/**
	 * Stops frame capture.
	 * <p>
	 * Behavior:
	 * <li>Drains or discards in-flight frames (implementation-dependent).</li>
	 * <li>Safe to call multiple times (idempotent).</li>
	 *
	 * @throws IllegalStateException if the instance is disposed.
	 */
	public native void stop();

	/**
	 * Releases native resources associated with this capture instance.
	 * <p>
	 * Lifecycle:
	 * <li>Implicitly stops capture if currently running.</li>
	 * <li>After disposal, further method calls (other than additional dispose attempts)
	 *     are invalid and may throw {@link IllegalStateException}.</li>
	 * <p>
	 * Best Practice:
	 * Always invoke in a finally block or use a higher-level resource management
	 * construct to avoid native leaks.
	 */
	public native void dispose();

	/**
	 * Internal native initialization hook invoked exactly once by the constructor.
	 * Not intended for direct external use.
	 * <p>
	 * Failure Handling:
	 * Should raise a runtime exception if initialization fails.
	 */
	private native void initialize();

}
